File: HackerCup/2014/R1/labelmaker.cpp
#include <iostream>
using namespace std;

#define REP(i,n) for(int i=0;i<(n);++i)
#define FOR(i,a,b) for(int i=(a);i<=(b);++i)
typedef long long LL;

void run() {
    string letters;
    LL n;
    cin >> letters >> n;
    int len = letters.length();
    string ret = "";
    while (n > 0) {
        ret = letters[--n % len] + ret;
        n /= len;
    }
    cout << ret << endl;
}

int main() {
    int kase;
    cin >> kase;
    FOR(k,1,kase) {
        cout << "Case #" << k << ": ";
        run();
    }
    return 0;
}

File: HackerCup/2014/R1/labelmaker.java
import java.util.*;
import java.io.*;
import java.math.*;

public class labelmaker
{    
    public static void main (String[] argv) throws IOException
	{
        BufferedReader in = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(in.readLine());
        int kase = Integer.parseInt(st.nextToken());
        for (int k = 1; k <= kase; ++k) {
			System.out.print(String.format("Case #%s: ", k));

			String letters;
			BigInteger n;
			st = new StringTokenizer(in.readLine());
			letters = st.nextToken();
            n = BigInteger.valueOf(Long.parseLong(st.nextToken()));
            
            int len = letters.length();
            if (len == 1) {
            	for (Long i = 0L; i < n.longValue(); ++i) {
            		System.out.print(letters);
            	}
            	System.out.println();
            	continue;
            }
            
            int sz = 0;
            BigInteger tmp = BigInteger.ONE;
            BigInteger bLen = BigInteger.valueOf(len);
            while (true) {
            	++sz;
            	tmp = tmp.multiply(bLen);
            	if (n.compareTo(tmp) <= 0) {
            		tmp = tmp.divide(bLen);
            		break;
            	}
            	n = n.subtract(tmp);
            }
            
            n = n.subtract(BigInteger.ONE);
            StringBuilder ret = new StringBuilder();
            for (int i = 0; i < sz; ++i) {
            	ret.append(letters.charAt(n.divide(tmp).intValue()));
            	n = n.mod(tmp);
            	tmp = tmp.divide(bLen);
            }
            
            System.out.println(ret.toString());
        }
    }
}

